src/NHibernate/Type/CollectionType.cs

```diff
@@ -359,36 +360,54 @@ public virtual object InstantiateResult(object original)
 			return result;
 		}
 
-		public virtual object ReplaceElements(object original, object target, object owner, IDictionary copyCache,
-											  ISessionImplementor session)
+		public virtual object ReplaceElements(object original, object target, object owner, IDictionary copyCache, ISessionImplementor session)
 		{
-			// TODO: does not work for EntityMode.DOM4J yet!
-			object result = target;
-			Clear(result);
+			var elemType = GetElementType(session.Factory);
+			var targetPc = target as IPersistentCollection;
+			var originalPc = original as IPersistentCollection;
+			var iter = (IEnumerable)original;
+			var clearTargetsDirtyFlag = ClearTargetsDirtyFlag(target, targetPc, originalPc, iter);
 
 			// copy elements into newly empty target collection
-			IType elemType = GetElementType(session.Factory);
-			IEnumerable iter = (IEnumerable)original;
-			foreach (object obj in iter)
+			Clear(target);
+			foreach (var obj in iter)
 			{
-				Add(result, elemType.Replace(obj, null, session, owner, copyCache));
+				Add(target, elemType.Replace(obj, null, session, owner, copyCache));
 			}
 
-			// if the original is a PersistentCollection, and that original
-			// was not flagged as dirty, then reset the target's dirty flag
-			// here after the copy operation.
-			// One thing to be careful of here is a "bare" original collection
-			// in which case we should never ever ever reset the dirty flag
-			// on the target because we simply do not know...
-			IPersistentCollection originalPc = original as IPersistentCollection;
-			IPersistentCollection resultPc = result as IPersistentCollection;
-			if(originalPc != null && resultPc!=null)
+			if(clearTargetsDirtyFlag)
+			{
+				targetPc.ClearDirty();
+			}
+
+			return target;
+		}
+
+		protected bool ClearTargetsDirtyFlag(object target, IPersistentCollection targetPc, IPersistentCollection originalPc, IEnumerable iter)
```

Comment:
I understand the need for this method, but I dislike its name and semantic. It sounds too internal for being exposed to users implementing their own collection type (normally by deriving from `CustomCollectionType`, which derives from `CollectionType`).

I think it should be `internal` instead of `protected`.